### DISCUSSION TITLE
2주차 멘토링

### DISCUSSION THREAD

**Discussion by arabae** [bcaitech1/p4-dkt-freshtomato#62] [General]
**멘토님의 생각**

정형데이터과 같이 느껴졌음

- pre-trained 모델을 쓴다고해서 좋아지는게 아니니까 feature-engineering이 더욱 중요한 것 같다고 생각

**멘토님의 2주차 추천**

1. 완전 다른 모델을 사용하되 feature를 괜찮은 것 같은 것을 사용하여 적용
2. binary다 보니까 DST만큼 output을 보고 분포를 보고 얻을 수 있는 것 확실히 적은 것 같음
    - 난이도를 낮추다보니 0/1만 check하는 것으로 생각이 듦
    - user들의 category를 분류했으면 좋겠음
    - 외부 다른 모델로 classification하여 user별로 맞춘 문제와 틀린 문제를 쌓다보니 clustering 등을 사용하여 생성한 결과를 다시 feature로 넣어주는 방식
    - kaggle에서 dataset에 대해서만 모델링하는게 한계이다보니 그래프 등을 가지고와서 classification, clustering을 사용하여 feature로 사용하는 방식이 필요한 것 같음
    - 새로운 것을 계속 활용해보는 사람과 기존의 모델에 feature를 추가하여 성능을 개선하는게 좋은 것 같음
    - 하나의 column당 10개의 feature 등을 만들다보면 무조건 성능이 오름..!! (마지막에 반드시 해봐야하는 것)
    - 잘 working하는지 계속 확인하는 건 중요한 것 같음
3. feature를 넣고 뺐을 때, 영향력을 확인 할 수는 있는 것 같음!
4. augmentation을 할 수 있는 task인가요?
    - (민용) 한 사람당 문제를 푼게 최소 9~1200개가 있어서 부분적으로 잘라서 사용하면 그런 효과를 줄 수 있는 것 같아요
5. 이번 주 내에 validation set을 나누는 기준을 확실히하면 좋을 것 같음
    - feature도 많이 추가하고, 다른 모델에도 적용해보고 있으니까 stratified를 다시 적용해보는게 좋을 것 같음

➕ 그래프를 활용한 clustering 후 이를 feature로도 사용

- transformer 등의 모델을 계속 활용했으면 좋겠음

실험 방향을 확고히 정해서 3주차때는 버리고 남아있는 중요한 애들을 사용해서 더 많은 실험을 하고, 4주차에는 앙상블만 계속하는게 타임라인 상 좋을 것 같음

- catboost, lgbm, xgboost를 앙상블해서 좋은지를 확인 → transfomer등과 비교해서 어느것이 더 좋을지 먼저 판단해야할 것 같음
- fold를 많이 나눴을 때 좋았음 (10개 정도)
- pseudo-labeling하기 clustering을 기반으로 lable을 정하는게 좋을 것 같음
- 지금 현재 SOTA를 확인하고, result 표를 보고 느낌을 받기!
- DKT ~ SAKT까지는 성능 차이가 별로 안났는데 SAINT에서 확 뛰었음, 이것을 확인하는게 좋을 것 같음
    - 중요한 feature 몇 개를 가지고 Multi-head를 사용한 transformer에 기대하는 느낌!
    - exercies id embedding, part embedding → 실제로 DKT에 중요한 embedding을 넣어서 사용하는 것 같음
    - DST에서 Q, K, V를 custom하게 사용해서 test한 사람도 있었음

✔️ 어려운 모델에 누군가 빠르게 시작하기

- DL 팀에서 transformer 류 모델을 사용해서 input feature를 다르게해서 사용해보는게 좋을 것 같음

categorical을 label/one-hot encoding 중 어떤거로 하나요?

- class number를 붙이는 것보다 **one-hot**이 나을 수도 있음
- 왜냐면 class가 증가될 때마다 좋아지는게 아니니까
- 정답 수는 더 커질수록 좋은 feature니까 괜찮은데, 아닌 것은 oen-hot이 나을 수 있음
- continuous 또한 분포를 나누어 categorical한 feature로도 추가할 수 있음!
    - 이렇게 했을 때 실제로 성능이 오른 경험도 있음

❓개수는 category인가요? continuous인가요?

- 숫자자체가 의미를 가진다면 discrete해보여도 continuous로 보는게 맞지 않을까요?

➕ 모델에 user의 개인 별 특징을 넣으면 좋을 것 같음

- 추천과 비슷하다는 생각이 들어서, 결국 맞추냐 마냐니까 추천과 같이 이 물건을 사냐 마냐와 같은 것 같음
- user의 embedding을 따로 구하고, 문제의 embedding을 따로 빼서 사용해보는 것도 좋을 것 같음
- 어떤 loss function이 좋을까? 생각해보기

**[ DL에서도 team 나눠서 여러가지로 실험해보기 ]**

1. feature의 영향 확인
2. 새로운 모델 적용

❓ 문제 별 정답률은 0~1사이 값만 가지고 있는데, 이 사람이 총 몇 문제를 풀었는지 feature는 1800도 최대값이 나오는데 이 값의 범위차가 영향이 있지 않나요?

너무 차이가 난다면 scale하는게 좋을 것 같음

- DL에서는 영향이 많지 않지만 0~1 / 0~1800 → scaling하면 좋아질 순 있음
- embedding 과정에서 상쇄될 수 있기때문에 많은 성능 변화가 일어나지 않아도 시도해보면 좋을 것 같음!

❓ category와 continue feature를 합치는 방법이 지금과 다르게 할 수 있는 방법이 있을까요?

단순히 concat은 가장 쉬운 방법이고, feature들을 multi-head attention으로 묶어서 넣어줄 수 있음

1. continue를 linear 전에 embedding해도 상관은 없음
    - 멘토님은 일단 실험해보고, 확인하는 편!
2. concat 전에 layer를 하나 더 두어 학습하고 더 좋은 feature로 만드는 방법?
3. layer norm하고 나온 vector를 multi-head attention를 사용해서 결합하는 방법도 있음
    - reference을 할 수 있는 데이터가 있는 경우가 있다면 더 쉬운데
4. 가중치를 둬서 concat

어떻게하면 concat해서 사용하면 좋은지, general하게 사용하는 방법이 뭐가 있는지 좀 더 찾아보고 알려주시겠다고 했음 🙂
